### PR TITLE
[mergebot][BE] If we don't have any required checks, don't run required checks

### PR DESCRIPTION
**Pull request by PaliC**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #121921
* #121920

This PR addresses the issue identified in #121920. The existing problem is that all tests are deemed mandatory if none are selected as required. This behavior is particularly noticeable during a force merge operation.

In the context of a force merge, it may not be necessary to execute any tests which are not required (imo). However, this proposed change could be seen as controversial, hence it has been separated from the main update for further discussion and review.

### CODE REVIEW

**Comment by huydhn** [comment]
After https://github.com/pytorch/pytorch/pull/121920, do we need this change to make force merge work with ET?  This `categorize_checks` function is only used to, well, categorize all the CI signals and shouldn't influence force merge.  

On the other hand, categorizing all signals when people use force merge is the expected behavior as explained in https://github.com/pytorch/pytorch/blob/main/.github/scripts/trymerge.py#L1443-L1445.  We want to know why people need to use force merge, be it failures (force merge due to failure) or pending (force merge due to impatience).  This behavior is actually powering the 2 metrics you see on https://hud.pytorch.org/metrics

**Comment by PaliC** [comment]
@huydhn  > After https://github.com/pytorch/pytorch/pull/121920, do we need this change to make force merge work with ET?
We definitely don't which is why I split it out :) 

Ahh I see. In practice though we likely just don't want to run all of the tests.
Though if we are force merging we should not be running any tests right? Anyways we end up running them do to the implementation detail here https://github.com/pytorch/pytorch/blob/53d2188df9930a5110ae286224cfd9867797acd8/.github/scripts/trymerge.py#L1996C1-L1996C78

Though in practice we can just short circuit if we are force merging and there are no required tests. I'll modify the PR for this. This just seems like a nonintuitive behavior to me to consider all the tests mandatory if none of them actually are.

File: .github/scripts/trymerge.py [hunk 1]
```diff
@@ -1411,6 +1411,9 @@ def find_matching_merge_rule(
             if rule.ignore_flaky_failures
             else 0,
         )
+        if len(required_checks) == 0:
```
Comment (huydhn) [inline .github/scripts/trymerge.py:1414]:
```suggestion
        if not required_checks
```

This should be sufficient :)

**Review by huydhn — APPROVED** [review]
LGTM! I only have a minor comment

**Pull request closed without merging**